src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        stateCopy = { ...stateCopy, ...action.extraData };
+        break;
+      }
+
+      case 'removeProperties': {
+        stateCopy = { ...stateCopy };
+
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+      }
+
+      case 'clear': {
+        stateCopy = {};
+        break;
+      }
+
+      default: {
+        throw new Error('Error!');
```

Comment:
The error message 'Error!' is not very descriptive. Consider providing a more informative message, such as 'Unknown action type: ' + action.type, to help with debugging.

Reply:
> The error message 'Error!' is not very descriptive. Consider providing a more informative message, such as 'Unknown action type: ' + action.type, to help with debugging.

Good point